### PR TITLE
67 resourceutilities types

### PR DESCRIPTION
**Pull request by TahaAttari**
Updated the types in `ResourceUtils` to make them safer to use for all FHIR versions. Now it is obvious what kind of input the utility is expecting.

### CODE REVIEW

**Comment by AD1306** [comment]
please provide tests for the changes you made to fix the pipeline.

**Comment by AD1306** [comment]
you also need to bump the version in package json

File: src/FHIR-DSTU2.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /* This is base class from which other elements are derived */
 export class FHIRElement {
-  id: string;
-  extension: Extension[];
+  id?: string;
+  extension?: Extension[];
```
Comment (TahaAttari) [inline src/FHIR-DSTU2.ts:4]: should be optional per the spec

File: src/FHIR-DSTU2.ts [hunk 2]
```diff
@@ -83,25 +83,25 @@ export class HumanName extends FHIRElement {
...
-  valueIdentifier: string;
+  valueHumanName?: HumanName;
+  valueReference?: Reference;
+  valueDate?: Date;
+  valueIdentifier?: string;
   valueDecimal?: number;
-  valueInteger: number;
-  valuePeriod: Period;
+  valueInteger?: number;
+  valuePeriod?: Period;
```
Comment (TahaAttari) [inline src/FHIR-DSTU2.ts:96]: all Extension values are optional per the spec

File: src/FHIR-R4/classes/extension.ts [hunk 3]
```diff
@@ -57,7 +57,7 @@ export class Extension {
     /**
      * Source of the definition for the extension code - a logical name or a URL.
      */
-    url?: Uri;
+    url: Uri;
```
Comment (TahaAttari) [inline src/FHIR-R4/classes/extension.ts:60]: url is mandatory per the spec

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,6 +1,11 @@
 # Changelog
 
-## 2.2.2
+## 2.2.0
```
Comment (AD1306) [inline CHANGELOG.md:3]: issue: the version is 2.3.0 per your package json.

File: src/library/ResourceUtils/ResourceUtils.ts [hunk 5]
```diff
@@ -1,3 +1,13 @@
+import * as DSTU2 from "../../FHIR-DSTU2";
+import * as R3 from "../../FHIR-R3";
+import { Coding as r4Coding } from "../../FHIR-R4/classes/coding";
+import { Extension as r4Extension } from "../../FHIR-R4/classes/extension";
+import { Identifier as r4Identifier } from "../../FHIR-R4/classes/identifier";
+export type Identifier = r4Identifier & R3.Identifier & DSTU2.Identifier;
+type IdentifierKeys = keyof Identifier;
+export type Extension = r4Extension & R3.Extension & DSTU2.Extension;
+export type Coding = r4Coding & R3.Coding & DSTU2.Coding;
+type CodingKeys = keyof Coding;
```
Comment (AD1306) [inline src/library/ResourceUtils/ResourceUtils.ts:10]: move all of this imports to a new file in its own folder under `library` folder, which only exports the types and keep this class as clean as possible. I'm pretty sure we'll use this types with other utils as well.

File: CHANGELOG.md (hunk 4 above)
Comment (TahaAttari) [reply CHANGELOG.md:3]: updated

File: src/library/ResourceUtils/ResourceUtils.ts (hunk 5 above)
Comment (TahaAttari) [reply src/library/ResourceUtils/ResourceUtils.ts]: moved

**Review by AD1306 — APPROVED** [review]

**Pull request merged**